keps/sig-autoscaling/20181106-in-place-update-of-pod-resources.md

```diff
@@ -0,0 +1,350 @@
+  denoting **desired** state of the Pod,
+* PodStatus.ContainerStatus.ResourceAllocated (new object) denotes **actual**
+  state of the Pod resources.
+
+To distinguish between possible states of a Pod resource update,
+a new PodCondition named ResourceResizeRequired is added, with the following states:
+* (empty) - the default value; resource update awaits reconciliation
+  if ResourceRequirements differs from ResourceAllocated,
+* Requested - Scheduler determined in-place resource resizing is possible, and
+  requested Kubelet to update Pod's resource allocations and limits,
+* Awaiting - awaiting resources to be freed (e.g. via pre-emption),
+* Failed - resource resizing could not have been performed in-place
+  but might be possible if some conditions change,
+* Rejected - resource update was rejected by any of the components involved.
+
+To provide some fine-grained control to the user,
+PodSpec.Container.ResourceRequirements is extended with ResizePolicy flag
+for each resource type (CPU, memory) :
+* NoRestart - the default value; resize the Container without restarting it,
+* RestartContainer - restart the Container in-place to apply new resource
+  values (e.g. Java process needs to change its Xmx flag),
+* RestartPod - restart the whole Pod in-place to apply new resource values
+  (e.g. Pod requires its Init Containers to re-run).
+
+By using the ResizePolicy flag, user can mark Containers or Pods as safe
+(or unsafe) for in-place resource update.
+
+This flag is used by Kubelet to determine the actions needed. This flag **may** be
+used by the actors starting the update to decide if the process should be started
+at all (for example VPA might decide to evict Pod with RestartPod policy).
+
+Setting the flag to separately control CPU & memory is due to an observation
+that usually CPU can be added/removed without much problems whereas
+changes to available memory are more probable to require restarts.
+
+If more than one resource type with different policies are updated, then
+RestartPod policy takes precedence over RestartContainer, which in turn takes
+precedence over NoRestart policy.
+
+#### CRI Changes
+
+Kubelet calls UpdateContainerResources CRI API which currently takes
+*runtimeapi.LinuxContainerResources* parameter that works for Docker and Kata,
+but not for Windows. This parameter is changed to *runtimeapi.ContainerResources*,
+that is runtime agnostic.
+
+### Flow Control
+
+The following steps denote a typical flow of an in-place resize process for a Pod
+with ResizePolicy set to Update for all its Containers.
+
+1. The initiating actor updates ResourceRequirements using PATCH verb.
+1. API Server validates the new ResourceRequirements
+   (e.g. limits are not below requested resources, QoS class does not change).
+1. API Server calls all Admission Controllers to verify the Pod Update.
+   * If any of the Controllers rejects the update, the
+     ResourceResizeRequired PodCondition is set to Rejected.
```

Comment:
This is different from most kubernetes behavior, which uses the retry-backoff pattern.  Can you give an example of where it would be beneficial to "stop trying" by setting it to rejected, rather than retrying on an interval?

Reply:
This came from the [above comment](https://github.com/kubernetes/enhancements/pull/686#discussion_r249052974) where the system is unable to do anything about it.

I can picture an unlikely example where user updates a controller template PodSpec, say asking for additional CPU for its pod instances, and it passes resource quota validation at that point. But say a scale up is in progress for that NS, and when the controller propagates the change to pod instances, we are over-budget. This would be a case for Rejected for the pod instance, as retrying wont help.

I'm working on policy spec for a case where retry would help - e.g. user wants in-place only and is willing to suffer sub-optimal resource allocation, but wants to get resized if other pods depart that node, and resources become available.

Reply:
I still don't quite understand.  Wouldn't the scale-up be rejected in that case?

In general, it seems like the way kubernetes handles "rejection events", such as failing to schedule it to omit an event, but not put that information in the status.

Reply:
No. I tried a simple experiment where I put a sleep in my job controller after validation succeeded and pod template spec updated, but just before propagating resize to the pod. During that time I scheduled a few RS and they successfully scheduled, using up RQ against NS. So when it gets to resource update, this would fail .. and we want to have a way to distinguish it.

We hit something similar while developing our first version sometime late last year, I don't recall the details now.

Reply:
I should be more specific.  One of the two should fail.  Either the step "propagating resize to the pod" should fail, or "schedule new pod as part of a new RS" should fail.  How can we end up over-budget?

Reply:
It will fail due to going over-budget on resource-quota. From @kgolab's comment, I think the idea is to distinguish this kind of failure from a failure due to node fit for example. In the latter case, VPA can choose to delete the pod to reschedule to a different node, whereas that action is of no use in the first case.

Reply:
Oh, ok.  I get it.  The issue isn't that we violate quota guarantees.  The issue is that we don't like it that our resize gets stuck indefinitely.  Will controllers always want to put the pod on a different node if it fails, or would that be a policy knob on VPA?  If we always want to move the pod to a different node, the controller that "permanently fails" the resize could instead just gracefully delete the pod, which would have the same effect.

Reply:
Not always, this should be a policy controlled choice. I updated the KEP adding [RetryPolicy](https://github.com/kubernetes/enhancements/pull/686/files#diff-57e345a188df5653422eb0ed20bf9442R158) to the API, hopefully the choices will address most scenarios users might have.

VPA, in this instance, would use the default policy and choose to delete the pod to reschedule it, alternately use Reschedule policy, or offer user a choice whether to retry in-place.

Reply:
We have dropped the event based retry for now, keeping backoff-retry typical to other parts of k8s. I've discussed handing RetryPolicy in VPA with sig-autoscaling. There are no open, unresolved items in this issue, so resolving it. Please reopen if not resolved.